torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -700,15 +835,17 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
         local_used_maps_reduced_ = true;
       }
 
-      auto bucket_view =
-          replica.contents.narrow(0, offset, length).view(variable.sizes());
+      const auto& bucket_view = replica.bucket_views[intra_bucket_index];
       runGradCallbackForVariable(variable, [&](auto& grad) {
         // If a parameter is globally unused, we keep its grad untouched.
         if (!global_unused) {
           if (!grad.defined()) {
-            grad = at::empty(bucket_view.sizes(), bucket_view.options());
+            // Creates grad according to the "Gradient Layout Contract"
```

Comment:
@mrshenli are the diffs here ok?  (again, grad might not be rowmajor contiguous here).  Also under what circumstances does the grad need to be "written back" and what does "written back" entail?  The grad is modified in place, so references to it elsewhere don't need to be explicitly modified (although I guess references in other processes would be)?

Reply:
Also, why isn't there an equivalent `runGradCallbackForVariable` in `finalize_bucket_sparse`?  Was that an oversight in the `dist_autograd` PR, or was that intentionally omitted?

Reply:
> are the diffs here ok?

This looks OK to me. 

> Also under what circumstances does the grad need to be "written back" and what does "written back" entail?

If you are referring to the `written back` comment below, I think that comment is inaccurate. IIUC, it should return true when we are certain that the grad value is final, and we can now launch cbs on it.

> Also, why isn't there an equivalent runGradCallbackForVariable in finalize_bucket_sparse? Was that an oversight in the dist_autograd PR, or was that intentionally omitted?

I guess this is because #37998 didn't mean to support sparse tensors for RPC + DDP. @pritamdamania87 if this is the case, should we also remove the `runGradCallbackForVariable` in `mark_variable_ready_sparse` to make it consistent?

Reply:
I think this was probably an oversight in #37998, we should probably call `runGradCallbackForVariable` in finalize_bucket_sparse as well. I'll fix this in a separate diff.

Reply:
SGTM.  I wrote my other thoughts about sparse grads in the other thread https://github.com/pytorch/pytorch/pull/34904/files#r439117597